[NUI] Removed binding and added command

zhouleonlei opened the pull request:
### Description of Change ###
1. Removed binding between style and view
2. Added Command in Control

dongsug-song commented on src/Tizen.NUI.Components/Controls/DropDown.cs:
```diff
@@ -205,9 +205,118 @@ public enum ListOrientation
+        /// </summary>
+        /// <since_tizen> 8 </since_tizen>
+        /// This will be public opened in tizen_6 after ACR done. Before ACR, need to be hidden as inhouse API.
+        public TextLabel HeaderText
```
Hello,
is it OK not to add [EditorBrowsable .. Never] ? 
I can see other APIs have this. 
also please remove <since_tizen> tag and remove tizen_6 word. (please change as tizen_next.)
thank you.

zhouleonlei replied:
The class of DropDown has been hidden. So there is not ACR notice of public api added. I will add [Never] also.

dongsug-song commented on src/Tizen.NUI.Components/Controls/DropDown.cs:
```diff
@@ -205,9 +205,118 @@ public enum ListOrientation
+        /// </summary>
+        /// <since_tizen> 8 </since_tizen>
+        /// This will be public opened in tizen_6 after ACR done. Before ACR, need to be hidden as inhouse API.
+        public Button Button
```
same as above.
Don't need [Editable .. Never] ?

zhouleonlei replied:
The class of DropDown has been hidden. So there is not ACR notice of public api added. I will add [Never] also.

dongsug-song commented on src/Tizen.NUI.Components/Controls/Loading.cs:
```diff
@@ -122,7 +125,13 @@ public Loading(LoadingStyle loadingStyle) : base(loadingStyle)
         /// </summary>
         /// <since_tizen> 8 </since_tizen>
-        public new LoadingStyle Style => ViewStyle as LoadingStyle;
+        public new LoadingStyle Style
```
Hello,
is this for only getting style?
I think we need to add some comment and explanation how style can be set.
maybe style setting is only possible by using constructor and ApplyStyleXXX() method?

rabbitfor replied:
Please add a comment that it returns copied Style instance, so that changing it does not effect to the corresponding View.

zhouleonlei replied:
OK. I will add

dongsug-song commented on src/Tizen.NUI.Components/Controls/Popup.cs:
```diff
@@ -331,7 +331,32 @@ public void AddContentText(View childView)
         /// </summary>
         /// <since_tizen> 8 </since_tizen>
-        public new PopupStyle Style => ViewStyle as PopupStyle;
+        //public new PopupStyle Style => ViewStyle as PopupStyle;
```
Hello, 
please remove this if it is useless

zhouleonlei replied:
Removed

dongsug-song commented on src/Tizen.NUI.Components/Controls/Popup.cs:
```diff
@@ -331,7 +331,32 @@ public void AddContentText(View childView)
+        }
+
+        [EditorBrowsable(EditorBrowsableState.Never)]
+        public TextLabel Title
```
Hello,
please add comments to explain what is Title.

dongsug-song commented on src/Tizen.NUI.Components/Controls/Scrollbar.cs:
```diff
@@ -184,7 +184,65 @@ public enum DirectionType
+        }
+
+        [EditorBrowsable(EditorBrowsableState.Never)]
+        public ImageView TrackImage
```
Hello,
please add comments to explain what is TrackImage

zhouleonlei replied:
OK I will add the comments

dongsug-song commented on src/Tizen.NUI.Components/Controls/Scrollbar.cs:
```diff
@@ -413,6 +459,24 @@ public void SetCurrentValue(int currentValue, bool enableAnimation = true)
+        /// <param name="viewStyle">The style to apply.</param>
+        /// <since_tizen> 8 </since_tizen>
+        [EditorBrowsable(EditorBrowsableState.Never)]
+        public override void ApplyStyle(ViewStyle viewStyle)
```
Hello, 
please remove <since_tizen> tag, if it is hidden API. (tagged with [Editor ... Never] )

zhouleonlei replied:
ok

dongsug-song commented on src/Tizen.NUI.Components/Controls/Switch.cs:
```diff
@@ -378,7 +406,7 @@ public ImageView GetCurrentTrack(SwitchExtension extension)
         public ImageView GetCurrentThumb(SwitchExtension extension)
         {
-            return (extension == Extension) ? Thumb : null;
+            return /*(extension == Extension) ? Thumb :*/ null;
```
Hello, 
please remove /* xxx */

zhouleonlei replied:
OK

rabbitfor replied:
Please remvoe both API `GetCurrentTrack` and `GetCurrentThumb`.
They are not needed anymore because now the Switch provides getter of Track and Thumb.

dongsug-song commented on src/Tizen.NUI.Components/Controls/Toast.cs:
```diff
@@ -79,13 +100,21 @@ public static Toast FromText(string text, uint duration)
         /// </summary>
         /// <since_tizen> 8 </since_tizen>
-        public new ToastStyle Style => ViewStyle as ToastStyle;
+        public new ToastStyle Style
```
Hello,
this is only getter.
I think some explanation needs to be added how user can set style.

zhouleonlei replied:
I  will add the comment

dongsug-song commented on src/Tizen.NUI.Components/PreloadStyle/OverlayAnimationButtonStyle.cs:
```diff
@@ -109,7 +109,7 @@ public override void OnControlStateChanged(Button button, ControlStates previous
             }
 
-            var overlayImage = button.GetCurrentOverlayImage(this);
+            ImageView overlayImage  = null/*= button.GetCurrentOverlayImage(this)*/;
```
please remove /* xxx */

zhouleonlei replied:
OK

rabbitfor replied:
ImageView overlayImage  = button.OverlayImage;

rabbitfor commented:
Hi Lei, 
Button Extension is needed to support Wearable Styles. (LottieButtonStyle, for example).
Can you revive ButtonExtension code back?

rabbitfor commented on src/Tizen.NUI.Components/Controls/Button.cs:
```diff
@@ -277,55 +243,117 @@ public enum IconOrientation
+                }
+                return buttonIcon;
+            }
+            set
```
How about changing setter to an internal API?

rabbitfor replied:
Or setter may need to reset like below,
```
Button.viewStyle.Icon = value.Style;
```

zhouleonlei replied:
Fixed

rabbitfor commented on src/Tizen.NUI/src/public/BaseComponents/View.cs:
```diff
@@ -2473,22 +2466,9 @@ private TriggerableSelector<Rectangle> backgroundImageBorderSelector
-            }
-
-            this.viewStyle = viewStyle;
+            this.ViewStyle.CopyFrom(viewStyle);
```
There is a case that a component getting extended style (LottieButtonStyle for a Button, for example),
so the code `this.ViewStyle.CopyFrom(viewStyle)` can lost extended data.
Please fix it like below code,
```
this.viewStyle =  viewStyle.Clone();
```

zhouleonlei replied:
Fixed

rabbitfor commented:
> Hi Lei,
> Button Extension is needed to support Wearable Styles. (LottieButtonStyle, for example).
> Can you revive ButtonExtension code back?

rabbitfor commented on src/Tizen.NUI.Components/Controls/Button.cs:
```diff
@@ -931,55 +965,12 @@ protected void UpdateState(Touch touchInfo = null)
-                ButtonText.ApplyStyle(style.Text);
-            }
-
+            iconRelativeOrientation = IconOrientation.Left; 
```
Button.ApplyStyle() is called before calling Button.Initialize().
So this over-write ButonStyle's IconOrientation value always.
Can you move this line to line297 or somewhere valid?

rabbitfor replied:
I found this issue by running this [sample](https://github.com/dalihub/nui-demo/tree/linux/wearable-samples/Button/ButtonWithOverlayAnimation).

You can check valid screen-shot in README.md.

zhouleonlei replied:
Fixed

rabbitfor commented:
For the properties that are selector type in Style,
they should be over-written when user sets corresponding View's property.

Button's BackgroundColor for example,
```
var style = new ButtonStyle
{
    BackgroundColor = new Selector<Color>
    {
        Normal = Color.Blue,
        Pressed = Color.Green,
    },
};

var button = new Button(style)
{
    Text = "Hello World!",
    Size = new Size(210, 72),
};

button.BackgroundColor = Color.Red;

window.Add(button);
```
Please run above example. You will see what I's trying to tell :)

AchoWang commented:
Looks good to me.

rabbitfor commented on src/Tizen.NUI.Components/Controls/Button.cs:
```diff
@@ -931,55 +965,11 @@ protected void UpdateState(Touch touchInfo = null)
-            {
-                Add(ButtonIcon);
-                ButtonIcon.ApplyStyle(style.Icon);
-                ButtonIcon.Relayout += OnIconRelayout;
```
listening Icon's relayout signal is needed.
In the first, Button does not know the correct size of the Icon. (image is not loaded yet)
 it can look weird in the first if you remove  this line.

Please run [this](https://github.com/dalihub/nui-demo/blob/linux/wearable-samples/Button/ButtonWithOverlayAnimation/README.md) example and do clicking what the difference.

rabbitfor commented:
And please update NUI TCT (include manua tc) not to use Style after create an instance.

zhouleonlei commented:
> Hello,
> please add new sample app or change tizenfx/test/Tizen.NUI.Samples when adding new APIs or many codes are changed.
Added it in https://github.com/Samsung/TizenFX/pull/1595

zhouleonlei commented:
> Hi Lei,
> Button Extension is needed to support Wearable Styles. (LottieButtonStyle, for example).
> Can you revive ButtonExtension code back?
> 
> We need not to make any regression of below examples,
> 
> * [ButtonWithOverlayAnimation](https://github.com/dalihub/nui-demo/tree/linux/wearable-samples/Button/ButtonWithOverlayAnimation)
> * [WearableCheckBox](https://github.com/dalihub/nui-demo/tree/linux/wearable-samples/Controls/WearableCheckBox) - Tizen only (not working in linux)
> * [WearableRadioButton](https://github.com/dalihub/nui-demo/tree/linux/wearable-samples/Controls/WearableRadioButton) - Tizen only (not working in linux)
> * [WearableSwitch](https://github.com/dalihub/nui-demo/tree/linux/wearable-samples/Controls/WearableSwitch) - Tizen only (not working in linux)

ButtonExtension has been revived

zhouleonlei commented:
> And please update NUI TCT (include manua tc) not to use Style after create an instance.

Has added Relayout for button

zhouleonlei commented:
> For the properties that are selector type in Style,
> they should be over-written when user sets corresponding View's property.
> 
> Button's BackgroundColor for example,
> 
> ```
> var style = new ButtonStyle
> {
>     BackgroundColor = new Selector<Color>
>     {
>         Normal = Color.Blue,
>         Pressed = Color.Green,
>     },
> };
> 
> var button = new Button(style)
> {
>     Size = new Size(210, 72),
> };
> 
> button.BackgroundColor = Color.Red;
> 
> window.Add(button);
> ```
> 
> Please run above example. You will see what I'm trying to tell :)

Fixed

zhouleonlei commented:
> @zhouleonlei
> Please keep one topic for one PR from next time. This PR seems to include at least 3 topics.
> 
> 1. Removed binding between style and view
> 2. Added Command in Control
> 3. Added layout and refine some codes
> 
> Please reply for all comments.

OK. Finished

hyunjushin commented:
Please check if this patch is still available.

AdunFang reviewed (commented):
It is OK.
